rotkehlchen/rotkehlchen.py

```diff
@@ -555,23 +556,82 @@ def get_history_query_status(self) -> Dict[str, str]:
 
     def process_history(
             self,
+            report_id: int,
             start_ts: Timestamp,
             end_ts: Timestamp,
     ) -> Tuple[Dict[str, Any], str]:
-        (
-            error_or_empty,
-            history,
-            loan_history,
-            asset_movements,
-            eth_transactions,
-            defi_events,
-            ledger_actions,
-        ) = self.events_historian.get_history(
-            start_ts=start_ts,
-            end_ts=end_ts,
-            has_premium=self.premium is not None,
-        )
+        def process_cache(cache_data: List[Dict[str, Any]]) -> Union[None, HistoryResult]:
+            if cache_data:
+                actions = [y for x in cache_data for y in x.values()]
+
+                return (
+                    error_or_empty,
+                    list(filter(lambda x: type(x) in ['Trade',
+                                                      'MarginPosition',
+                                                      'AMMTrade'], actions)),
+                    list(filter(lambda x: type(x) in ['Loan'], actions)),
+                    list(filter(lambda x: type(x) in ['AssetMovement'], actions)),
+                    list(filter(lambda x: type(x) in ['EthereumTransaction'],
+                                actions)),
+                    list(filter(lambda x: type(x) in ['DefiEvent'], actions)),
+                    list(filter(lambda x: type(x) in ['LedgerAction'], actions)))
```

Comment:
here you are iterating the same list a gazillion times. Just iterate the cache data in a god old for loop and during that one iteration create all the relevant needed sublists.